File: c++/2. Add Two Numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode solution = new ListNode(0);
        ListNode solutionStart = solution;
        int carry = 0;

        while (true) {
            int x;
            if (l1 != null)
                x= l1.val;
            else
                x = 0;

            int y;
            if (l2 != null )
                y=l2.val;
            else
                y = 0;

            if ((x + y + carry) < 10) {
                solution.val = x + y + carry;
                carry = 0;
            } else { 
                solution.val = (x + y + carry) % 10;
                carry = 1;
            }

            //move pointers to the next segment
            if (l1 != null)
                l1 = l1.next;

            if (l2 != null)
                l2 = l2.next;

            if ((l1 == null) && (l2 == null) && (carry == 0)) {
                //no more elements left to add
                break;
            } else {
                solution.next = new ListNode(0);
                solution = solution.next; //traverse to next segment
            }

        
        } //while 
        return solutionStart; //return reference to the first node   
    }
}

File: c++/2. Add Two Numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        int carry = 0;
        ListNode *solution = new ListNode(0); 
        ListNode *returnPtr = solution;

        while(true){
            int x = l1 ? l1 -> val : 0;
            int y = l2 ? l2 -> val : 0;
            int sum = x + y + carry;
            if (sum < 10) {
                solution->val = sum;
                carry = 0;
            } else {
                solution->val = sum %10;
                carry = 1;
            }
            
            //traverse to the next node on the linked lists
            l1 =  l1 ? l1->next : nullptr;
            l2 =  l2 ? l2->next : nullptr;

            if (l1 == NULL && l2 == NULL && carry == 0)
                break;
            else
                solution->next = new ListNode(0);
                solution = solution -> next;

        } //while
        return returnPtr;
    }
};